File: 프로그래머스/2/42885. 구명보트/구명보트.py
# 1. 한 번에 최대 2명씩 밖에 탈 수 없고, 무게 제한도 있다.
# 2.구명보트를 최대한 적게 사용하여 모든 사람 구출
# 3. 모든 사람을 구출하기 위해 필요한 구명보트 개수의 최솟값 return
# left, right로 나누어서 구하는 방법

def solution(people, limit):

    people.sort()
    answer = 0
    
    # 여기서 left 는 0, right는 left가 0부터 시작하니 len(people)-1
    left = 0 
    right = len(people) - 1  

    while left <= right:
        if people[left] + people[right] <= limit:
            left += 1  
        
        # 두 명을 태울 수 없는 경우, 가장 무거운 사람은 혼자라도 무조건 보트에 타야 함
        right -= 1
        answer += 1  # 보트를 하나 사용합니다.
    
    return answer

File: 프로그래머스/2/42885. 구명보트/구명보트.java
import java.util.Arrays;

class Solution {
    // 구명보트를 최대한 적게 사용해 모든 사람을 구출
    public int solution(int[] people, int limit){
        // 몸무게를 오름차순으로 정렬
        Arrays.sort(people);
        // 필요한 보트 개수
        int count = 0;
        // 가장 가벼운 사람을 가리키는 인덱스
        int i = 0;
        // 가장 무거운 사람을 가르키는 인덱스
        int j = people.length -1;
        
        while (i<=j){
            // 가장 무거운 사람과 가장 가벼운 사람을 같이 태울 수 있으면 두 사람 모두 보트에 태움
            if (people[i] + people[j] <= limit){
                i += 1;
            }
            // 무거운 사람만 태울 수 있으면 무거운 사람만 보트에 태움
            j -= 1; 
            count += 1;
        }
        return count;
    }
}